vcs/util/tracer/tracer.go

```diff
@@ -0,0 +1,132 @@
+package tracer
+
+import (
+	"fmt"
+	"time"
+
+	"sourcegraph.com/sourcegraph/appdash"
+	"sourcegraph.com/sourcegraph/go-vcs/vcs"
+)
+
+func init() { appdash.RegisterEvent(GoVCS{}) }
+
+// GoVCS records a go-vcs method invocation.
+type GoVCS struct {
+	Name, Args string
+
+	StartTime time.Time
+	EndTime   time.Time
+}
+
+// Schema returns the constant "GoVCS".
+func (GoVCS) Schema() string { return "GoVCS" }
+
+func (e GoVCS) Start() time.Time { return e.StartTime }
+func (e GoVCS) End() time.Time   { return e.EndTime }
+
+// Wrap wraps the given VCS repository, returning a repository which emits
+// tracing events.
+func Wrap(r vcs.Repository, rec *appdash.Recorder) vcs.Repository {
+	// Wrap the repository.
+	t := repository{r: r, rec: rec}
+
+	// Also wrap optional interface.
+	final := vcs.Repository(t)
+	if b, ok := r.(vcs.Blamer); ok {
+		final = blamer{Repository: final, b: b, rec: rec}
+	}
+	if d, ok := r.(vcs.Differ); ok {
+		final = differ{Repository: final, d: d, rec: rec}
+	}
+	if c, ok := r.(vcs.CrossRepoDiffer); ok {
+		final = crossRepoDiffer{Repository: final, c: c, rec: rec}
+	}
+	if f, ok := r.(vcs.FileLister); ok {
+		final = fileLister{Repository: final, f: f, rec: rec}
+	}
+	return final
+}
+
+// blamer implements the vcs.Repository interface, adding a wrapped vcs.Blamer
+// implementation.
+type blamer struct {
+	vcs.Repository
```

Comment:
Wouldn't this be more robust if the struct embedded the local `repository` struct directly? That way, you'd insure that that all the `vcs.Repository` methods were properly wrapped.